### PR TITLE
[XPU][Diffusion] Enable MiniMax H3 on XPU platforms

### PR DESCRIPTION
**Pull request by jianan-gu**
This PR adds key changes to cover the enabling for Minimax H3 model specific on XPU platforms, including

- Mapping to XPU attention backend with optimized kernel from sgl-kernel-xpu
- Necessary device dispatching/autocast codes
- Runtime XPU memory D2H/H2D fix

Tested serving pipelines include:

- TP + layer_wise offload for both encoder/dit/vae ( similar as cookbook on cuda 5090: https://docs.sglang.io/cookbook/diffusion/MiniMax/MiniMax-H3#3-serve-minimax-h3, but preferring no offload (only encoder) when device memory is enough, i.e., x4 or x8) 

Next plans (as follow-up PRs, for further optimizations).

- Apply more fusions (like QKnorm_RoPE and silu_and_mul, to be on par with cuda fusion pathes)
- More serving pipelines/parallelism to be explored and evaluated.
- XPU FA backend with non-contiguous support

Tested example:

- T2VA example from cookbook: https://docs.sglang.io/cookbook/diffusion/MiniMax/MiniMax-H3#4-generate-video-and-audio

Result:

https://github.com/user-attachments/assets/09e086a7-95ae-489b-bb9b-5e48553eca91

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #34183716549](https://github.com/sgl-project/sglang/actions/runs/34183716549)
Latest PR Test (Extra): :white_check_mark: [Run #34183716399](https://github.com/sgl-project/sglang/actions/runs/34183716399)
Latest PR Test (AMD ROCm 7.2): :x: [Run #34183716525](https://github.com/sgl-project/sglang/actions/runs/34183716525)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py [hunk 1]
```diff
@@ -284,10 +284,75 @@ def _silu_mul(hidden: torch.Tensor, *, reuse_input: bool) -> torch.Tensor:
         and hidden.shape[-1] % 16 == 0
     ):
         return silu_and_mul_with_activation_rounding_(hidden)
+    if (
```
Comment (mickqian) [inline python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py]: could we move this either to platform or kernel file?

Comment (jianan-gu) [reply python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py]: Sure, thanks for the suggestion, will refine this part . (and all similar ones).

Comment (jianan-gu) [reply python/sglang/multimodal_gen/runtime/models/dits/minimax_h3.py]: Hi, regarding to those applying more fusions, as in PR description, we are removing them in follow-up PRs, as we need more time to catch up full fusion patterns as cuda does. And it can help make this PR more clear.

**Comment by jianan-gu** [comment]
@mickqian  Hi~, would you help revisit and review this PR? Thanks!

**Comment by jianan-gu** [comment]
The only CI failure also exists on main : https://github.com/sgl-project/sglang/actions/runs/31059443986

**Comment by jianan-gu** [comment]
Hi @mickqian, soft reminder, any chance to review this PR changes? Thanks!

File: python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py [hunk 2]
```diff
@@ -120,3 +120,30 @@ def forward(
 
         result = out.reshape(bsz, seqlen_q, nheads_q, d)
         return result
+
+    def forward_varlen(
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py:124]: i guess this was covered by `python/sglang/multimodal_gen/test/unit/test_minimax_h3_dit_contract.py`?

File: python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py [hunk 3]
```diff
@@ -120,3 +120,30 @@ def forward(
...
+        value: torch.Tensor,
+        *,
+        cu_seqlens: torch.Tensor,
+        max_seqlen: int,
+        cu_seqlens_host: tuple[int, ...] | None = None,
+    ) -> torch.Tensor:
+        del cu_seqlens_host
+        q_ = query.contiguous()
+        k_ = key.contiguous()
+        v_ = value.contiguous()
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py:137]:
```suggestion
```

`flash_attn_func` should support lastdim contiguous q k v.

File: python/sglang/multimodal_gen/runtime/managers/memory_managers/component_manager.py [hunk 4]
```diff
@@ -447,9 +447,11 @@ def _finish_use(
         if should_keep:
             return
         strategy = self.strategy_for(use.component_name, module)
-        was_on_cuda = self._module_on_cuda(module)
+        was_on_local_accelerator = self._module_on_local_accelerator(module)
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/managers/memory_managers/component_manager.py]: here you name it after `local_accelerator` because `local_device` includes cpu?

https://github.com/sgl-project/sglang/pull/33366/changes#diff-bc08c727df87288952ecca3818540d7defa373024d31b105efbd7a90c9515515R28-R39

File: python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py (hunk 3 above)
Comment (jianan-gu) [reply python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py:137]: Actually not, we ran into accuracy issue without contiguous.
And previous integration also adds contuguous: https://github.com/sgl-project/sglang/pull/33366/changes/BASE..62e85fe263128b85ef31313aecd2aee6a04da908#diff-3b19ab55fff1c6de62e11b1b55c067d99975df13db26ea996dbf08d5c5c4b19bR97-R99
We may further improve here later, and list it in next plans to follow up.

File: python/sglang/multimodal_gen/runtime/managers/memory_managers/component_manager.py (hunk 4 above)
Comment (jianan-gu) [reply python/sglang/multimodal_gen/runtime/managers/memory_managers/component_manager.py]: yes, local device is from current_platform which includes cpu: https://github.com/jianan-gu/sglang/blob/62e85fe263128b85ef31313aecd2aee6a04da908/python/sglang/multimodal_gen/runtime/distributed/group_coordinator.py#L48

File: python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py (hunk 2 above)
Comment (jianan-gu) [reply python/sglang/multimodal_gen/runtime/layers/attention/backends/xpu_backend.py:124]: checked it only covers SDPA fallback test.

File: python/sglang/multimodal_gen/runtime/models/encoders/qwen3vl.py [hunk 5]
```diff
@@ -590,6 +590,8 @@ def forward(
         hidden_states = inputs_embeds
 
         # create position embeddings to be shared across the decoder layers
+        position_ids = position_ids.to(hidden_states.device)
+        self.rotary_emb.to(hidden_states.device)
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/models/encoders/qwen3vl.py]: cuda is OK without these two lines?

File: python/sglang/multimodal_gen/runtime/managers/memory_managers/component_resident_strategies.py [hunk 6]
```diff
@@ -446,9 +446,9 @@ def enter(self, module: nn.Module) -> None:
             _module_to_local_device(module)
 
     def exit(self, module: nn.Module, next_module: nn.Module | None = None) -> None:
-        param = next(module.parameters(), None)
-        if param is not None and param.device.type == "cuda":
-            module.to("cpu", non_blocking=True)
+        tensor = _module_reference_tensor(module)
+        if tensor is not None and tensor.device.type != "cpu":
+            module.to("cpu", non_blocking=tensor.device.type == "cuda")
```
Comment (mingfeima) [inline python/sglang/multimodal_gen/runtime/managers/memory_managers/component_resident_strategies.py]: CUDA behavior is not changed.
NPU/MUSA/MPS previously skipped D2H and would now do a blocking .to("cpu"). 

check on device.type in ("cuda", "xpu") (or current_platform.device_type)

**Review by mingfeima — APPROVED** [review]
some minor changes needed.

File: python/sglang/multimodal_gen/runtime/managers/memory_managers/component_resident_strategies.py (hunk 6 above)
Comment (jianan-gu) [reply python/sglang/multimodal_gen/runtime/managers/memory_managers/component_resident_strategies.py]: Sure, have refined here accordingly with in ("cuda", "xpu")

File: python/sglang/multimodal_gen/runtime/models/encoders/qwen3vl.py (hunk 5 above)
Comment (jianan-gu) [reply python/sglang/multimodal_gen/runtime/models/encoders/qwen3vl.py]: Have double checked here with latest base, XPU is same as cuda, for cpu_offload/layerwise_offload and even FSDP, the buffers are already on device. Thus removing those two lines as no need.

**Comment by jianan-gu** [comment]
Rebased, CI passed: https://github.com/sgl-project/sglang/actions/runs/31766407552/job/94663260263?pr=33366
Will try to re-run AMD suite.

**Comment by mingfeima** [comment]
@jianan-gu the xpu ci skipped due to https://github.com/sgl-project/sglang/blob/main/.github/workflows/pr-test-xpu.yml#L57

@arathi-hlab any particular reason that `multimodal_gen` folder is skipped?

**Comment by arathi-hlab** [comment]
https://github.com/sgl-project/sglang/pull/21482 its skipped by radix community itself , if you want we can enable it

**Comment by jianan-gu** [comment]
Hi @mickqian could you help revisit and take a look on this PR? Thanks.
The PR changes much simpler now as we have rebased with main that like NPUs are doing similar refactoring for H3.
The CI is passed for CUDA, and the AMD CI test failures are not related to this PR, also happens on main: https://github.com/sgl-project/sglang/actions/runs/33456144533

**Comment by jianan-gu** [comment]
Hi @mingfeima, after rebase from main, this PR is XPU specific now, all previous common components changes now we are leveraging the common path from main, by just adding `is_xpu`. And the main CI (CUDA) is passed, with some failures (AMD suite) also happen on main. Shall we proceed? Thanks

**Comment by mingfeima** [comment]
@jianan-gu rebase and fix the conflict.
